pkg/schedule/coordinator.go

```diff
@@ -152,72 +149,195 @@ func (c *Coordinator) IsPendingRegion(region uint64) bool {
+}
+
+func (p *PatrolRegionContext) init(ctx context.Context, cluster sche.ClusterInformer) {
+	p.interval = cluster.GetCheckerConfig().GetPatrolRegionInterval()
+	p.workerCount = cluster.GetCheckerConfig().GetPatrolRegionWorkerCount()
+
+	p.scanLimit = calculateScanLimit(cluster)
+	p.patrolRoundStartTime = time.Now()
+
+	p.regionChan = make(chan *core.RegionInfo, patrolRegionChanLen)
+	p.workersCtx, p.workersCancel = context.WithCancel(ctx)
+}
+
+func (p *PatrolRegionContext) stop() {
+	close(p.regionChan)
+	p.workersCancel()
+	p.wg.Wait()
+}
+
+func (p *PatrolRegionContext) updateScanLimit(cluster sche.ClusterInformer) {
+	p.scanLimit = calculateScanLimit(cluster)
+}
+
+func calculateScanLimit(cluster sche.ClusterInformer) int {
+	return max(patrolScanRegionMinLimit, cluster.GetTotalRegionCount()/patrolRegionPartition)
+}
+
+func (p *PatrolRegionContext) getPatrolRegionsDuration() time.Duration {
+	p.durationLock.RLock()
+	defer p.durationLock.RUnlock()
+	return p.duration
+}
+
+func (p *PatrolRegionContext) setPatrolRegionsDuration(dur time.Duration) {
+	p.durationLock.Lock()
+	defer p.durationLock.Unlock()
+	p.duration = dur
+}
+
+func (p *PatrolRegionContext) startPatrolRegionWorkers(c *Coordinator) {
+	for i := 0; i < p.workerCount; i++ {
+		p.wg.Add(1)
+		go func() {
+			defer logutil.LogPanic()
+			defer p.wg.Done()
+			for {
+				select {
+				case region, ok := <-p.regionChan:
+					if ok {
+						c.tryAddOperators(region)
+					} else {
+						log.Debug("region channel is closed")
+					}
+				case <-p.workersCtx.Done():
+					log.Debug("region worker is closed")
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (p *PatrolRegionContext) roundUpdate() {
+	dur := time.Since(p.patrolRoundStartTime)
+	patrolCheckRegionsGauge.Set(dur.Seconds())
+	p.setPatrolRegionsDuration(dur)
+	p.patrolRoundStartTime = time.Now()
+}
+
 // PatrolRegions is used to scan regions.
 // The checkers will check these regions to decide if they need to do some operations.
 // The function is exposed for test purpose.
 func (c *Coordinator) PatrolRegions() {
 	defer logutil.LogPanic()
-
 	defer c.wg.Done()
-	ticker := time.NewTicker(c.cluster.GetCheckerConfig().GetPatrolRegionInterval())
+
+	ctx, cancel := context.WithCancel(c.ctx)
+	defer cancel()
+	c.patrolRegionContext.init(ctx, c.cluster)
+	defer c.patrolRegionContext.stop()
+
+	ticker := time.NewTicker(c.patrolRegionContext.interval)
 	defer ticker.Stop()
 
+	c.patrolRegionContext.startPatrolRegionWorkers(c)
+
 	log.Info("coordinator starts patrol regions")
-	start := time.Now()
 	var (
 		key     []byte
 		regions []*core.RegionInfo
 	)
 	for {
 		select {
 		case <-ticker.C:
-			// Note: we reset the ticker here to support updating configuration dynamically.
-			ticker.Reset(c.cluster.GetCheckerConfig().GetPatrolRegionInterval())
+			c.updateTickerIfNeeded(ticker)
+			c.updatePatrolWorkersIfNeeded()
+
+			if c.cluster.IsSchedulingHalted() {
+				for len(c.patrolRegionContext.regionChan) > 0 {
```

Comment:
What about adding a log to indicate how many regions dropped here?

Reply:
I think it seems to be not necessary